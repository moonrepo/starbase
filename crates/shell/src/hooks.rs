pub enum Statement<'data> {
    PrependPath {
        paths: &'data [String],
        key: Option<&'data str>,
        orig_key: Option<&'data str>,
    },
    SetEnv {
        key: &'data str,
        value: &'data str,
    },
    UnsetEnv {
        key: &'data str,
    },
}

pub enum Hook {
    OnChangeDir { command: String, prefix: String },
}

impl Hook {
    pub fn get_info(&self) -> &str {
        match self {
            Hook::OnChangeDir { .. } => "on change directory",
        }
    }
}
