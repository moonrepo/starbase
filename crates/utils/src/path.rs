use std::char::REPLACEMENT_CHARACTER;
use std::ffi::OsStr;
use std::hash::{DefaultHasher, Hasher};
use std::path::{Component, Path, PathBuf};

/// Normalize separators in a path string to their OS specific separators.
/// On Unix this will be `/`, and on Windows `\`.
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    #[cfg(unix)]
    {
        path.as_ref().replace('\\', "/")
    }

    #[cfg(windows)]
    {
        path.as_ref().replace('/', "\\")
    }
}

/// Standardize separators in a path string to `/` for portability,
#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

/// Format the provided name for use as an executable file.
/// On Windows this will append `.exe`, on Unix used as-is.
#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    let name = name.as_ref();

    #[cfg(unix)]
    {
        name.into()
    }

    #[cfg(windows)]
    {
        if name.ends_with(".exe") {
            name.into()
        } else {
            format!("{name}.exe")
        }
    }
}

/// Encode a value by removing invalid characters for use within a path component.
pub fn encode_component<T: AsRef<OsStr>>(value: T) -> String {
    let mut output = String::new();

    for ch in value.as_ref().to_string_lossy().chars() {
        match ch {
            '@' | '*' | REPLACEMENT_CHARACTER => {
                // Skip these
            }
            '/' | ':' => {
                output.push('-');
            }
            _ => {
                output.push(ch);
            }
        }
    }

    output.trim_matches(['-', '.']).to_owned()
}

/// Hash a value that may contain special characters into a valid path component.
pub fn hash_component<T: AsRef<OsStr>>(value: T) -> String {
    let mut hasher = DefaultHasher::default();
    hasher.write(value.as_ref().as_encoded_bytes());

    format!("{}", hasher.finish())
}

/// Clean a path by removing and flattening unnecessary path components.
pub fn clean<T: AsRef<Path>>(path: T) -> PathBuf {
    // Based on https://gitlab.com/foo-jin/clean-path
    let mut components = path.as_ref().components().peekable();

    let mut cleaned = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
        components.next();

        PathBuf::from(c.as_os_str())
    } else {
        PathBuf::new()
    };

    let mut leading_parent_dots = 0;
    let mut component_count = 0;

    for component in components {
        match component {
            Component::Prefix(_) | Component::CurDir => {}
            Component::RootDir => {
                cleaned.push(component.as_os_str());
                component_count += 1;
            }
            Component::ParentDir => {
                if component_count == 1 && cleaned.is_absolute() {
                    // Nothing
                } else if component_count == leading_parent_dots {
                    cleaned.push("..");
                    leading_parent_dots += 1;
                    component_count += 1;
                } else {
                    cleaned.pop();
                    component_count -= 1;
                }
            }
            Component::Normal(c) => {
                cleaned.push(c);
                component_count += 1;
            }
        }
    }

    if component_count == 0 {
        cleaned.push(".");
    }

    cleaned
}

/// Return true if both provided paths are equal. Both paths will be cleaned
/// before comparison for accurate matching.
pub fn are_equal<L: AsRef<Path>, R: AsRef<Path>>(left: L, right: R) -> bool {
    clean(left) == clean(right)
}

/// Extend the native [`Path`] and [`PathBuf`] with additional functionality.
pub trait PathExt {
    /// Clean a path by removing and flattening unnecessary path components.
    fn clean(&self) -> PathBuf;

    /// Return true if the current path matches the provided path.
    /// Both paths will be cleaned before comparison for accurate matching.
    fn matches(&self, other: &Path) -> bool;
}

impl PathExt for Path {
    fn clean(&self) -> PathBuf {
        clean(self)
    }

    fn matches(&self, other: &Path) -> bool {
        are_equal(self, other)
    }
}

impl PathExt for PathBuf {
    fn clean(&self) -> PathBuf {
        clean(self)
    }

    fn matches(&self, other: &Path) -> bool {
        are_equal(self, other)
    }
}
