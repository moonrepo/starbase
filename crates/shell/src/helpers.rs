use std::env;
use std::ffi::OsString;
use std::path::{Path, PathBuf};

#[cfg(unix)]
pub static PATH_DELIMITER: &str = ":";

#[cfg(windows)]
pub static PATH_DELIMITER: &str = ";";

pub fn is_absolute_dir(value: OsString) -> Option<PathBuf> {
    let dir = PathBuf::from(&value);

    if !value.is_empty() && dir.is_absolute() {
        Some(dir)
    } else {
        None
    }
}

pub fn get_config_dir(home_dir: &Path) -> PathBuf {
    env::var_os("XDG_CONFIG_HOME")
        .and_then(is_absolute_dir)
        .unwrap_or_else(|| home_dir.join(".config"))
}

pub fn get_env_var_regex() -> regex::Regex {
    regex::Regex::new(r"\$(?<name>[A-Z0-9_]+)").unwrap()
}

pub fn normalize_newlines(content: impl AsRef<str>) -> String {
    let content = content.as_ref().trim();

    #[cfg(windows)]
    {
        content.replace('\r', "").replace('\n', "\r\n")
    }

    #[cfg(unix)]
    {
        content.replace('\r', "")
    }
}
