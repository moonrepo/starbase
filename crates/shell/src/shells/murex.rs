use super::Shell;
use crate::helpers::{normalize_newlines, PATH_DELIMITER};
use crate::hooks::*;
use std::fmt;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug)]
pub struct Murex;

impl Murex {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        Self
    }
}

impl Shell for Murex {
    fn format(&self, statement: Statement<'_>) -> String {
        match statement {
            Statement::PrependPath {
                paths,
                key,
                orig_key,
            } => {
                let key = key.unwrap_or("PATH");
                let orig_key = orig_key.unwrap_or(key);

                format!(
                    r#"$ENV.{key}="{}{}$ENV.{orig_key}""#,
                    paths.join(PATH_DELIMITER),
                    PATH_DELIMITER,
                )
            }
            Statement::SetEnv { key, value } => {
                format!("$ENV.{}={}", self.quote(key), self.quote(value))
            }
            Statement::UnsetEnv { key } => {
                format!("unset {};", self.quote(key))
            }
        }
    }

    // hook referenced from https://github.com/direnv/direnv/blob/ff451a860b31f176d252c410b43d7803ec0f8b23/internal/cmd/shell_murex.go#L12
    fn format_hook(&self, hook: Hook) -> Result<String, crate::ShellError> {
        Ok(normalize_newlines(match hook {
            Hook::OnChangeDir { command, prefix } => {
                format!(
                    r#"
$ENV.__ORIG_PATH="$ENV.PATH"

event: onPrompt {prefix}_hook=before {{
  {command} -> source
}}
"#
                )
            }
        }))
    }

    fn get_config_path(&self, home_dir: &Path) -> PathBuf {
        home_dir.join(".murex_profile")
    }

    fn get_env_path(&self, home_dir: &Path) -> PathBuf {
        home_dir.join(".murex_preload")
    }

    fn get_profile_paths(&self, home_dir: &Path) -> Vec<PathBuf> {
        vec![
            home_dir.join(".murex_preload"),
            home_dir.join(".murex_profile"),
        ]
    }

    /// Quotes a string according to Murex shell quoting rules.
    /// @see <https://murex.rocks/tour.html#basic-syntax>
    fn quote(&self, value: &str) -> String {
        if value.starts_with('$') {
            return format!("\"{}\"", value);
        }

        // Check for simple values that don't need quoting
        if value
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-')
        {
            return value.to_string();
        }

        // Handle brace quotes %(...)
        if value.starts_with("%(") && value.ends_with(')') {
            return value.to_string(); // Return as-is for brace quotes
        }

        // Check for values with spaces or special characters requiring double quotes
        if value.contains(' ') || value.contains('"') || value.contains('$') {
            // Escape existing backslashes and double quotes
            let escaped_value = value.replace('\\', "\\\\").replace('"', "\\\"");
            return format!("\"{}\"", escaped_value);
        }

        // Default case for complex values
        value.to_string()
    }
}

impl fmt::Display for Murex {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "murex")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    #[test]
    fn formats_env_var() {
        assert_eq!(
            Murex.format_env_set("PROTO_HOME", "$HOME/.proto"),
            r#"$ENV.PROTO_HOME="$HOME/.proto""#
        );
    }

    #[cfg(unix)]
    #[test]
    fn formats_path() {
        assert_eq!(
            Murex.format_path_set(&["$PROTO_HOME/shims".into(), "$PROTO_HOME/bin".into()]),
            r#"$ENV.PATH="$PROTO_HOME/shims:$PROTO_HOME/bin:$ENV.PATH""#
        );
    }

    #[cfg(windows)]
    #[test]
    fn formats_path() {
        assert_eq!(
            Murex.format_path_set(&["$PROTO_HOME/shims".into(), "$PROTO_HOME/bin".into()]),
            r#"$ENV.PATH="$PROTO_HOME/shims;$PROTO_HOME/bin;$ENV.PATH""#
        );
    }

    #[test]
    fn formats_cd_hook() {
        let hook = Hook::OnChangeDir {
            command: "starbase hook murex".into(),
            prefix: "starbase".into(),
        };

        assert_snapshot!(Murex.format_hook(hook).unwrap());
    }

    #[test]
    fn test_murex_quoting() {
        assert_eq!(Murex.quote("value"), "value");
        assert_eq!(Murex.quote("value with spaces"), r#""value with spaces""#);
        assert_eq!(Murex.quote("$(echo hello)"), "\"$(echo hello)\"");
        assert_eq!(Murex.quote(""), "");
        assert_eq!(Murex.quote("abc123"), "abc123");
        assert_eq!(Murex.quote("%(Bob)"), "%(Bob)");
        assert_eq!(Murex.quote("%(hello world)"), "%(hello world)");
    }
}
