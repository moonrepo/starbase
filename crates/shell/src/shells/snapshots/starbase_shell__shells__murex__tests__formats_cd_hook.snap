---
source: crates/shell/src/shells/murex.rs
expression: Murex.format_hook(hook).unwrap()
---
$ENV.__ORIG_PATH="$ENV.PATH"

function _starbase_hook {
  starbase hook murex -> source
}

event onPrompt starbase_hook=before {
  _starbase_hook
}
